examples/ls-files.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * libgit2 "ls-files" example - shows how to view all files currently in the index
+ *
+ * Written by the libgit2 contributors
+ *
+ * To the extent possible under law, the author(s) have dedicated all copyright
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * You should have received a copy of the CC0 Public Domain Dedication along
+ * with this software. If not, see
+ * <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+
+#include "common.h"
+#include "array.h"
+
+/**
+ * This example demonstrates the libgit2 index APIs to roughly
+ * simulate the output of `git ls-files`.
+ * `git ls-files` has many options and this currently does not show them.
+ *
+ * `git ls-files` base command shows all paths in the index at that time.
+ * This includes staged and committed files, but unstaged files will not display.
+ *
+ * This currently supports:
+ * 	- The --error-unmatch paramter with the same output as the git cli
+ *  - default ls-files behavior
+ *
+ * This currently does not support:
+ * 	- anything else
+ *
+ */
+
+typedef struct {
+	int error_unmatch;
+	git_array_t(char *) files;
+} ls_options;
+
+/* Print a usage message for the program. */
+static void usage(const char *message, const char *arg)
+{
+	if (message && arg)
+		fprintf(stderr, "%s: %s\n", message, arg);
+	else if (message)
+		fprintf(stderr, "%s\n", message);
+	fprintf(stderr, "usage: ls-files [--error-unmatch] [--] [<file>...]\n");
+	exit(1);
+}
+
+static int parse_options(ls_options *opts, int argc, char *argv[])
+{
+	int parsing_files = 0;
+	struct args_info args = ARGS_INFO_INIT;
+	char **file;
+
+	memset(opts, 0, sizeof(ls_options));
+	git_array_init(opts->files);
+
+	if (argc < 2)
+		return 0;
+
+	for (args.pos = 1; args.pos < argc; ++args.pos) {
+		char *a = argv[args.pos];
+
+		/* if it doesn't start with a '-' or is after the '--' then it is a file */
+		if (a[0] != '-') {
+			parsing_files = 1;
+
+			file = git_array_alloc(opts->files);
+			GITERR_CHECK_ALLOC(file);
+			*file = a;
+		} else if (!strcmp(a, "--")) {
+			parsing_files = 1;
+		} else if (!strcmp(a, "--error-unmatch") && !parsing_files) {
```

Comment:
If you change the first condition as proposed, there's no need for the `!parsing_files` anymore.